# India Air Quality Analysis & Prediction

## 1. Project Overview


## 2. Dataset

In [11]:
import pandas as pd
import os

df = pd.read_csv(os.path.join('air_quality_index_dataset', 'city_day.csv'))
print(df.head(5))
print(df.shape)

        City        Date  PM2.5  PM10     NO    NO2    NOx  NH3     CO    SO2  \
0  Ahmedabad  2015-01-01    NaN   NaN   0.92  18.22  17.15  NaN   0.92  27.64   
1  Ahmedabad  2015-01-02    NaN   NaN   0.97  15.69  16.46  NaN   0.97  24.55   
2  Ahmedabad  2015-01-03    NaN   NaN  17.40  19.30  29.70  NaN  17.40  29.07   
3  Ahmedabad  2015-01-04    NaN   NaN   1.70  18.48  17.97  NaN   1.70  18.59   
4  Ahmedabad  2015-01-05    NaN   NaN  22.10  21.42  37.76  NaN  22.10  39.33   

       O3  Benzene  Toluene  Xylene  AQI AQI_Bucket  
0  133.36     0.00     0.02    0.00  NaN        NaN  
1   34.06     3.68     5.50    3.77  NaN        NaN  
2   30.70     6.80    16.40    2.25  NaN        NaN  
3   36.08     4.43    10.14    1.00  NaN        NaN  
4   39.31     7.01    18.89    2.78  NaN        NaN  
(29531, 16)


In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 29531 entries, 0 to 29530
Data columns (total 16 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   City        29531 non-null  object 
 1   Date        29531 non-null  object 
 2   PM2.5       24933 non-null  float64
 3   PM10        18391 non-null  float64
 4   NO          25949 non-null  float64
 5   NO2         25946 non-null  float64
 6   NOx         25346 non-null  float64
 7   NH3         19203 non-null  float64
 8   CO          27472 non-null  float64
 9   SO2         25677 non-null  float64
 10  O3          25509 non-null  float64
 11  Benzene     23908 non-null  float64
 12  Toluene     21490 non-null  float64
 13  Xylene      11422 non-null  float64
 14  AQI         24850 non-null  float64
 15  AQI_Bucket  24850 non-null  object 
dtypes: float64(13), object(3)
memory usage: 3.6+ MB


In [7]:
print(df.describe())

              PM2.5          PM10            NO           NO2           NOx  \
count  24933.000000  18391.000000  25949.000000  25946.000000  25346.000000   
mean      67.450578    118.127103     17.574730     28.560659     32.309123   
std       64.661449     90.605110     22.785846     24.474746     31.646011   
min        0.040000      0.010000      0.020000      0.010000      0.000000   
25%       28.820000     56.255000      5.630000     11.750000     12.820000   
50%       48.570000     95.680000      9.890000     21.690000     23.520000   
75%       80.590000    149.745000     19.950000     37.620000     40.127500   
max      949.990000   1000.000000    390.680000    362.210000    467.630000   

                NH3            CO           SO2            O3       Benzene  \
count  19203.000000  27472.000000  25677.000000  25509.000000  23908.000000   
mean      23.483476      2.248598     14.531977     34.491430      3.280840   
std       25.684275      6.962884     18.133775    

### Initial Observations

The dataset contains **29,531 records across 16 columns**, including city, date, pollutant concentrations, AQI, and AQI bucket.

Initial inspection shows that:
- `City` contains no missing values and can be used to identify locations.
- `Date` contains no missing values but is currently stored as a string and will need to be converted to datetime for time-based analysis.
- Several pollutant measurements contain missing values, with `PM10`, `NH3`, and `Xylene` having substantial gaps.
- The dataset contains both numerical pollutant measurements and categorical variables such as `City` and `AQI_Bucket`.
- `AQI` has **24,850 non-null observations**, leaving 4,681 records without an AQI value.
- The pollutant variables show considerable variation in their distributions. For example, `PM2.5` has a median of **48.57** compared with a maximum of **949.99**, while AQI has a median of **118** and a maximum of **2,049**.
- The large differences between median and maximum values suggest the presence of **extreme observations (outliers)**, which should be considered during exploratory analysis and modeling.


These observations will guide the data-cleaning and preprocessing steps in the next section.

## 3. Data Cleaning & Preprocessing

In [8]:
# Selecting columns to use 
cols_to_use = ['City', 'Date', 'PM2.5', 'PM10', 'NO2', 'AQI', 'AQI_Bucket']
df_extracted = df[cols_to_use].copy()
print(df_extracted.head())


        City        Date  PM2.5  PM10    NO2  AQI AQI_Bucket
0  Ahmedabad  2015-01-01    NaN   NaN  18.22  NaN        NaN
1  Ahmedabad  2015-01-02    NaN   NaN  15.69  NaN        NaN
2  Ahmedabad  2015-01-03    NaN   NaN  19.30  NaN        NaN
3  Ahmedabad  2015-01-04    NaN   NaN  18.48  NaN        NaN
4  Ahmedabad  2015-01-05    NaN   NaN  21.42  NaN        NaN


In [ ]:
# Handling Nulls
print(df_extracted.isnull().sum())

# Dropping rows where AQI is Null
df_extracted = df_extracted.dropna(subset = ['AQI'])
print(df_extracted.isnull().sum())
print(df_extracted.shape)

# Filling missing values of PM2.5 with its mean
df_extracted['PM2.5'] = df_extracted['PM2.5'].fillna(df_extracted['PM2.5'].mean())
print(df_extracted.isnull().sum())
print(df_extracted.shape)


City             0
Date             0
PM2.5          678
PM10          7086
NO2            391
AQI              0
AQI_Bucket       0
dtype: int64
City             0
Date             0
PM2.5          678
PM10          7086
NO2            391
AQI              0
AQI_Bucket       0
dtype: int64
(24850, 7)
City             0
Date             0
PM2.5            0
PM10          7086
NO2            391
AQI              0
AQI_Bucket       0
dtype: int64
(24850, 7)


In [ ]:
# Sorting
print(df_extracted.sort_values(by = ['City', 'AQI'], ascending = [True, False]))

            City        Date       PM2.5   PM10     NO2     AQI AQI_Bucket
1145   Ahmedabad  2018-02-19  242.660000    NaN  199.17  2049.0     Severe
1136   Ahmedabad  2018-02-10  185.770000    NaN  172.84  1917.0     Severe
560    Ahmedabad  2016-07-14   38.130000    NaN   48.06  1842.0     Severe
1048   Ahmedabad  2017-11-14  207.190000    NaN  146.18  1747.0     Severe
1463   Ahmedabad  2019-01-03  131.500000    NaN   75.82  1719.0     Severe
...          ...         ...         ...    ...     ...     ...        ...
26330    Talcher  2018-10-14   67.476613   0.13   70.90    15.0       Good
25991   Shillong  2020-05-22    8.580000  12.38    2.84    14.0       Good
26308    Talcher  2018-09-22   67.476613   0.04   31.52    14.0       Good
25810   Shillong  2019-11-23    6.560000   8.70    0.95    14.0       Good
26329    Talcher  2018-10-13   67.476613   0.07   51.49    13.0       Good

[24850 rows x 7 columns]
                City        Date       PM2.5   PM10     NO2     AQI AQI_Bu

In [14]:
# Aggregating
print(df_extracted.groupby('City')['AQI'].mean().round(2))

City
Ahmedabad             452.12
Aizawl                 34.77
Amaravati              95.30
Amritsar              119.92
Bengaluru              94.32
Bhopal                132.83
Brajrajnagar          150.28
Chandigarh             96.50
Chennai               114.50
Coimbatore             73.02
Delhi                 259.49
Ernakulam              92.36
Gurugram              225.12
Guwahati              140.11
Hyderabad             109.21
Jaipur                133.68
Jorapokhar            159.25
Kochi                 104.28
Kolkata               140.57
Lucknow               217.97
Mumbai                105.35
Patna                 240.78
Shillong               53.80
Talcher               172.89
Thiruvananthapuram     75.88
Visakhapatnam         117.27
Name: AQI, dtype: float64


In [15]:
print(df_extracted.groupby('AQI_Bucket').size())

AQI_Bucket
Good            1341
Moderate        8829
Poor            2781
Satisfactory    8224
Severe          1338
Very Poor       2337
dtype: int64


### Aggregation Insights

The AQI distribution shows that **Moderate** and **Satisfactory** readings make up the majority of observations, while **Very Poor** and **Severe** readings represent a substantial portion of the dataset.

At the city level, the mean AQI varies considerably across locations, with **Ahmedabad showing the highest mean AQI** and **Aizwal the lowest** in this dataset.

These differences motivate further analysis of AQI patterns by category and season.

### AQI Categorization

The dataset includes an `AQI_Bucket` column containing categorical AQI labels, but the numerical boundaries corresponding to these categories are not explicitly defined in the dataset.

To enable analysis using the numerical `AQI` values, we create a separate categorical variable by dividing AQI into defined ranges. These ranges will allow us to examine the distribution of air-quality conditions more systematically.

In [16]:
bins = [0, 50, 100, 200, 300, 500]
label_names = ["Good", "Moderate", "Poor", "Very Poor", "Severe"]
df_extracted_cut = pd.cut(df_extracted['AQI'], bins, labels = label_names)
print(df_extracted_cut.value_counts())


AQI
Poor         8829
Moderate     8224
Severe       3132
Very Poor    2781
Good         1341
Name: count, dtype: int64


### Comparing Derived and Original AQI Categories

The derived categories were compared with the original `AQI_Bucket` column to understand how the chosen numerical boundaries affect the results.

The comparison revealed important differences:

- The original `AQI_Bucket` contains **six categories**, including `Satisfactory`, while the custom categorization initially used five categories.
- As a result, observations that would fall into the `Satisfactory` range were grouped into `Moderate` in the custom categorization.
- The `Severe` category also differs substantially between the two approaches because the selected numerical boundaries cover a different AQI range.
- This demonstrates that **AQI category counts depend directly on the numerical boundaries chosen for each bin**.

For this analysis, the original `AQI_Bucket` is therefore retained as the source categorical variable, while the custom categorization is treated as a derived feature for analysis.

### Seasonal Classification

The date information is used (after converting to datetime and extracting months) to assign each observation to a season. This allows us to investigate whether air-quality conditions vary systematically across different seasons.

In [19]:
df_extracted['Date'] = pd.to_datetime(df_extracted['Date'], errors = 'coerce')
df_extracted['month_num'] = df_extracted['Date'].dt.month
print(df_extracted.head())

         City       Date   PM2.5  PM10    NO2    AQI AQI_Bucket  month_num  \
28  Ahmedabad 2015-01-29   83.13   NaN  28.71  209.0       Poor          1   
29  Ahmedabad 2015-01-30   79.84   NaN  28.68  328.0  Very Poor          1   
30  Ahmedabad 2015-01-31   94.52   NaN  32.66  514.0     Severe          1   
31  Ahmedabad 2015-02-01  135.99   NaN  42.08  782.0     Severe          2   
32  Ahmedabad 2015-02-02  178.33   NaN  35.31  914.0     Severe          2   

    Season  
28  Winter  
29  Winter  
30  Winter  
31  Winter  
32  Winter  


In [20]:
# Now mapping month numbers to seasons:
seasons_by_month = {
    1: "Winter", 2: "Winter", 12: "Winter",
    3: "Summer", 4: "Summer", 5: "Summer",
    6: "Monsoon", 7: "Monsoon", 8: "Monsoon", 9: "Monsoon",
    10: "Post-Monsoon", 11: "Post-Monsoon"
}
df_extracted['Season'] = df_extracted['month_num'].map(seasons_by_month)
print(df_extracted.head())
print(df_extracted.shape)

         City       Date   PM2.5  PM10    NO2    AQI AQI_Bucket  month_num  \
28  Ahmedabad 2015-01-29   83.13   NaN  28.71  209.0       Poor          1   
29  Ahmedabad 2015-01-30   79.84   NaN  28.68  328.0  Very Poor          1   
30  Ahmedabad 2015-01-31   94.52   NaN  32.66  514.0     Severe          1   
31  Ahmedabad 2015-02-01  135.99   NaN  42.08  782.0     Severe          2   
32  Ahmedabad 2015-02-02  178.33   NaN  35.31  914.0     Severe          2   

    Season  
28  Winter  
29  Winter  
30  Winter  
31  Winter  
32  Winter  
(24850, 9)


In [21]:
# Seasonal AQI Variation
df_seasonal = df_extracted.groupby('Season')['AQI'].mean().round(2)
print(df_seasonal)
print(df_seasonal.shape)


Season
Monsoon         115.56
Post-Monsoon    215.47
Summer          147.67
Winter          220.61
Name: AQI, dtype: float64
(4,)


### Seasonal AQI Patterns

The mean AQI varies substantially across seasons:

- **Winter:** 220.61
- **Post-Monsoon:** 215.47
- **Summer:** 147.67
- **Monsoon:** 115.56

Winter and post-monsoon periods show the highest mean AQI, while monsoon has the lowest mean AQI in this dataset.

This seasonal variation suggests that temporal and environmental factors may be relevant when analyzing air-quality patterns. Further analysis would be required to determine which specific climate variables contribute to these differences.

## 4. Exploratory Data Analysis

## 5. Data Visualization

## 6. Feature Engineering

## 7. AQI Regression

## 8. AQI Classification

## 9. Model Evaluation

## 10. Time-Series Analysis

## 11. Key Findings

## 12. Conclusion